### DISCUSSION TITLE
Setting Up PiKaraoke in Kiosk Mode on Raspberry Pi with Bookworm Lite OS

### DISCUSSION THREAD

**Discussion by lvmasterrj** [vicwomg/pikaraoke#430] [Show and tell]
Inspired by the discussion started by [@romeo1984 about running pikaraoke kiosk mode with FullPageOS](https://github.com/vicwomg/pikaraoke/discussions/406), I decided to write this tutorial on running pikaraoke in kiosk mode with the "native" Bookworm Lite OS.

My issue was that using the RPI OS with desktop on my Pi 3b+ was making the system too slow and the easy solution was to overclocking my Pi which I didn't wanted to, so I found this solution on running the chromium without the need of the desktop which consumes less resources.

Is important to note that I'm using a Pi 3b+, I don't see why it wouldn't work with other pi versions, but I can't test them.

First I installed the Bookworm Lite OS on my SD card, using the [Raspberry Pi Imager](https://www.raspberrypi.com/software/).

_In the Imager, you can configure the Wi-Fi SSID, password, and SSH keys. If you're unfamiliar with this, just select the OS image and proceed. We'll configure Wi-Fi directly on the Pi later._

SD card with image, plug it on the pi and start it on.

**Configuring Wi-Fi:**
After the boot process, run `nmtui` This will open the Network Manager where you can go to "Activate a connection", select your Wi-Fi and activate it. After that, exit the manager and let's start running some commands.

### Steps

**1. Updating the repositories:**
It's always a good practice to update your repositories.

In the CLI, run:
```
sudo apt update
sudo apt upgrade -y
```

**3. Installing the necessary componentes:**
Now we will install the necessary stuff to run the PiKaraoke and to run it in kiosk mode.

In the CLI, run:
```
sudo apt-get install ffmpeg -y # Necessary to run the videos
sudo apt-get install --no-install-recommends xserver-xorg -y # The window system
sudo apt-get install --no-install-recommends xinit -y # Necessary to start the window
sudo apt-get install --no-install-recommends chromium-browser -y # The browser to run the pikaraoke
sudo apt-get install --no-install-recommends chromium-chromedriver -y # Some importante browser things
sudo apt-get install openbox -y # A window manager
```

**4. Installing Pikaraoke**
Now we create and activate the virtual environment and install PiKaraoke

Create a .venv directory in the home directory:
```
python -m venv ~/.venv
```

Activate your virtual environment:
```
source ~/.venv/bin/activate
```

Install PiKaraoke from PyPi:
```
pip install pikaraoke
```

**5. Configuring Kiosk Mode**
In this step, we will configure the RPi to start the right things to run in kiosk mode

Create a .xinitrc file in the home directory:
```
nano ~/.xinitrc
```

Add the following to launch the browser in kiosk mode
```
#!/bin/sh
xset -dpms  # disable DPMS (Energy Star) features
xset s off  # disable screen saver
xset s noblank  # don’t blank the video device
. ~/.venv/bin/activate
exec openbox-session & pikaraoke
```

**Creating a Systemd Service to Start the System**
Create a new service file called startx.service in the /etc/systemd/system/ directory:
```
sudo nano /etc/systemd/system/startx.service
```

Add the following content to the file:
```
[Unit]
Description=Start X at boot
After=network.target

[Service]
User=pi
Group=pi
Type=simple
Environment=DISPLAY=:0
Environment=XAUTHORITY=/home/pi/.Xauthority
ExecStart=/usr/bin/startx
WorkingDirectory=/home/pi
StandardInput=tty
StandardOutput=journal
StandardError=journal
TTYPath=/dev/tty1

[Install]
WantedBy=multi-user.target
```

Reload the systemd services to recognize the new service:
```
sudo systemctl daemon-reload
```

Enable the service so that it starts automatically on boot:
```
sudo systemctl enable startx.service
```

And thats it!

### One more thing
If you want to change the audio output, you can try the following:

1. In the CLI run:
```
aplay -l
```
It will list the playback hardware devices and show something like this:
![image](https://github.com/user-attachments/assets/837e1724-83cd-425c-ae9b-7a771ebe75c1)
Based on this, you will look for the desired output. In my case:

HEADPHONE - it is card 0
HDMI - it is card 1

Now, in the CLI, run:
```
sudo nano ~/.asoundrc
```

clear everything (or comment out the lines with a `#`) and put this in place, changing 0 to the desired output card number:
```
defaults.pcm.card 0
defaults.ctl.card 0
```

That's it. Hope it helps someone out there.

@vicwomg keep up the excellent work. Let's have fun!

**Comment by vicwomg** [comment]
Excellent guide! Really good to know that this helps performance on the pi 3. Did not realize the desktop introduces that much CPU overhead.

I will add to the wiki.

**Comment by dublus** [comment]
Thank you for making this tutorial. I tried it a week or so ago and was sure I followed all steps quite closely. It didn't work. When everything was done it just booted into the regular prompt. I used the 32 bit Bookworm Lite OS. Should I have used the 64 bit? I couldn't think of anything else that would be different from what you did.

**Reply by lvmasterrj** [reply]
Hello @dublus, sorry to hear that.

Mine is running on 64 bit. I can't tell if this is the problem, sorry.

Maybe it's a permission problem, let's try the following:

- run `sudo raspi-config`
- go to `System Options >> Boot / Auto Login >> Console Autologin` and reboot.

Let's see if it works.

If don't, try using a fresh 64 bit Bookworm Lite OS, and let me update.

**Reply by tabzz73** [reply]
Tried that too using 64bit, sometimes it boots in prompt, but reboot with ssh it goes through, another issue was that no audio from mp4 files, even download from youtube, but the audio from background music can be heard.

**Comment by dublus** [comment]
Thank you for the quick reply. I had already setup Auto Login. I've got a fresh 64 bit Bookworm Lite OS ready to give another try. There was one other place in the instructions I wasn't completely sure about. There was a point where there were further steps to follow and the prompt was the venv prompt. Was I supposed to exit out of that or execute the commands using that prompt?

**Comment by lvmasterrj** [comment]
No, you should run them in venv.
Hoppe it works this time. 🤞

**Comment by dublus** [comment]
OK, I just remembered something else. It seemed like the user/password should be pi and pikaraoke. Is that correct? It doesn't seem like that is specifically noted anywhere.

**Comment by Romeo1984** [comment]
Negative. That was way back when vicwomg was building the whole disk image. Today the user is whatever you make it. The default credentials are pi/raspberry for all of the standard raspian OS builds. Sent from my iPhoneOn Dec 6, 2024, at 2:36 PM, dublus ***@***.***> wrote:﻿
OK, I just remembered something else. It seemed like the user/password should be pi and pikaraoke. Is that correct? It doesn't seem like that is specifically noted anywhere.

**Comment by dublus** [comment]
OK, this part seemed confusing:

[Service]
User=pi

Since I don't have a pi user. There would be no path:

 /home/pi/ 

or am I missing something?

**Reply by lvmasterrj** [reply]
You are correct. I've used the default raspberry pi imager to send the OS to my sd and leaved the settings as default (username: pi).
I guess you have to change the service, where it is set to pi, to your user:

```
[Service]
**User=pi
**Group=pi
Type=simple
Environment=DISPLAY=:0
**Environment=XAUTHORITY=/home/pi/.Xauthority**
ExecStart=/usr/bin/startx
**WorkingDirectory=/home/pi
StandardInput=tty
StandardOutput=journal
StandardError=journal
TTYPath=/dev/tty1
```

That's the only thing that comes to my mind. About the password, it's not noted anywere in my instructions because it's really not needed.

Update me on your success.

**Comment by dublus** [comment]
Well, still no luck. Just boots into the normal prompt. As a last resort I'll ask a couple things that could have gone wrong on my end.

I'm seeing a space after the first "." .
. ~/.venv/bin/activate
Is that correct?

When you say "Create a .venv directory in the home directory:" and "Create a .xinitrc file in the home directory:" do you literally mean in the "home" directory? Because even if I'm in the "home" directory it always puts it in the "/home/pi" directory.

**Reply by lvmasterrj** [reply]
Sorry to hear that.

Regard your questions, the first "." is a short form for "source", and so the command is correct as it's like `source venv_path`
About the home directory, I mean `/home/pi`, just as you made it. I said home directory as it's the entry directory when you're logged in.

I'm realy sad that you're struggling to make it work, I know the frustation because before I wrote this guide I was struggling too. I wrote down every thing that work for me and only posted it here when I could reproduce every step and was successfull.
I will try to follow my tutorial step-by-step to see if I'm missing something.

**Reply by lvmasterrj** [reply]
To try to help, I used chatgpt to see if it enlight's us. Try these:

**1. Check the startx.service**
Ensure that the service was created correctly and is active.

Check the service status:
`sudo systemctl status startx.service`
_Expected Output: It should indicate the service is active (running). If there are errors, note them._

If the service is not running, try starting it manually:
_sudo systemctl start startx.service_

If errors occur, review the service file to ensure its content matches the tutorial:
`sudo nano /etc/systemd/system/startx.service`

**2. Check the .xinitrc**
Ensure the ~/.xinitrc file contains the correct configurations and is in the right location.

Open the file to verify:
`nano ~/.xinitrc`
Expected Content:

```
#!/bin/sh
xset -dpms  # disable DPMS (Energy Star) features
xset s off  # disable screen saver
xset s noblank  # don’t blank the video device
. ~/.venv/bin/activate
exec openbox-session & pikaraoke
```

Ensure the file has executable permissions:
`chmod +x ~/.xinitrc`

**3. Verify the Virtual Environment**
Confirm that the virtual environment exists and pikaraoke is installed.

Activate the virtual environment:
`source ~/.venv/bin/activate`

Check if pikaraoke is installed:
`pip list | grep pikaraoke`
_Expected Output: It should list pikaraoke along with its version._

**4. Test Startx Manually**
Try starting the graphical environment manually to see if errors occur.

Run the command:
**startx**
_If the graphical environment fails to start or errors appear, note the messages displayed._

**5. Check Logs for Errors**
Examine system logs to identify potential issues.

General system logs:
`journalctl -xe`

Logs specific to the service:
`journalctl -u startx.service`

**6. Other Possible Causes**

- Wrong user in the service: The startx.service file specifies the pi user. Ensure you’re using the correct username. If you’re using a different user, replace pi with your username in the service file.

- Missing packages: Ensure all necessary packages are installed:
`sudo apt-get install ffmpeg xserver-xorg xinit chromium-browser chromium-chromedriver openbox -y`

- Display permissions: You may need to set the display permissions:
`export DISPLAY=:0`

**Comment by Romeo1984** [comment]
Any thoughts on building on the FullPageOS? Others had success using my wiki.Sent from my iPhoneOn Dec 7, 2024, at 4:01 AM, Leonardo Moreira ***@***.***> wrote:﻿
To try to help, I used chatgpt to see if it enlight's us. Try these:
1. Check the startx.service
Ensure that the service was created correctly and is active.
Check the service status:
sudo systemctl status startx.service
Expected Output: It should indicate the service is active (running). If there are errors, note them.
If the service is not running, try starting it manually:
sudo systemctl start startx.service
If errors occur, review the service file to ensure its content matches the tutorial:
sudo nano /etc/systemd/system/startx.service
2. Check the .xinitrc
Ensure the ~/.xinitrc file contains the correct configurations and is in the right location.
Open the file to verify:
nano ~/.xinitrc
Expected Content:
#!/bin/sh
xset -dpms  # disable DPMS (Energy Star) features
xset s off  # disable screen saver
xset s noblank  # don’t blank the video device
. ~/.venv/bin/activate
exec openbox-session & pikaraoke

Ensure the file has executable permissions:
chmod +x ~/.xinitrc
3. Verify the Virtual Environment
Confirm that the virtual environment exists and pikaraoke is installed.
Activate the virtual environment:
source ~/.venv/bin/activate
Check if pikaraoke is installed:
pip list | grep pikaraoke
Expected Output: It should list pikaraoke along with its version.
4. Test Startx Manually
Try starting the graphical environment manually to see if errors occur.
Run the command:
startx
If the graphical environment fails to start or errors appear, note the messages displayed.
5. Check Logs for Errors
Examine system logs to identify potential issues.
General system logs:
journalctl -xe
Logs specific to the service:
journalctl -u startx.service
6. Other Possible Causes


Wrong user in the service: The startx.service file specifies the pi user. Ensure you’re using the correct username. If you’re using a different user, replace pi with your username in the service file.


Missing packages: Ensure all necessary packages are installed:
sudo apt-get install ffmpeg xserver-xorg xinit chromium-browser chromium-chromedriver openbox -y


Display permissions: You may need to set the display permissions:
export DISPLAY=:0

**Comment by dublus** [comment]
It's working now!

I have a feeling the problem was something different each time but in my sleep deprived state last night I made a typo on "WorkingDirectory" and instead had "Workingdirectory". That's the only thing I found and after I changed it it worked.

That's the good news. The bad news is there's something weird about the key changer. If I change the key the song stops as usual to start over with the new key but it never starts again. Just shows the song as the "Up next" song in the queue and in the home screen it says "Nothing is playing". In the admin interface it shows the CPU at 100%.

That reminds me that I should have asked about the key changer from the beginning. Do you have it working without any stutter? If I run Bookworm on my Pi3+ in desktop OS the song can't play smoothly once I change the key. I was hoping that running the lite OS might help.

**Reply by Hartebee5t** [reply]
Regarding the stuttering, I suspect it's that the Pi3+ is still not powerful enough, esp with desktop still behind the pikaraoke app. Key change/transpose is massively processor-hungry as it's having to transpose in real-time along with all the video streaming.

That prompts a question for the wider group - Could transpose be selected on demand by the host and set to be performed offline in slow time (however long it takes, similar to the file originally being downloaded through yt_dlp). This would create a new file in the library in the changed key that just runs like a fresh downloaded one without the realtime overhead, and therefore no stuttering or warbling artefacts?  Maybe I need to raise this in a new feature request...

**Reply by vicwomg** [reply]
Give the latest 1.8.0 a try. It has a transcode buffer size setting, which can help with pi 3 performance. I set it to 900kb and seems to be fine with keeping up with transcoding, but takes a bit longer to start up the song. Another cool feature: most settings are available from the info screen and don't need to be specified via the command line.

**Comment by vicwomg** [comment]
Personally not having trouble on my Pi3 using stock raspberry pi OS
bookworm desktop 64. No overlocking and the video is a bit choppy, but fine
for karaoke. With overlocking it’s a little better.

Not 100% convinced that running lite helps all that much.

**Reply by lvmasterrj** [reply]
Just curious, what are your overclocking settings?
I'm gonna give it a try.

**Comment by vicwomg** [comment]
I think I just used the settings from this thread:
https://forums.raspberrypi.com/viewtopic.php?t=235753

**Comment by bdmarinho** [comment]
First of all, thanks for this clear tutorial. It worked for me.
But I have an issue with video playback. Everything loads fine, but when video starts I have a pink backgroud with normal sound, QR code, IP, song time and  info.
Somebody has faced this issue?

**Reply by jimmerson17** [reply]
I'm getting this too.  Raspberry Pi 3b.  Fresh install of Pi OS Lite.

**Reply by safepay** [reply]
The pink screen is caused due to a lack of GPU processing of the browser window and videos. This also causes choppy video playback. You are actually better of using the desktop that comes with Bookworm because it uses the hardware more effectively. Bypassing the desktop actually reduces performance.

There have been many changes recently to the UI that speed up PiKaraoke on RPi, such as a lot more Ajax functions for browsing pages, etc. These ad up to a smoother overall experience.

However, I have updated the guide to allow Chrome to work without the desktop environment that includes the flag `--disable-gpu-compositing` the prevents the pink screen. But I do not recommend running PiKaraoke on a RPi like this.
